compiler/rustc_const_eval/src/interpret/eval_context.rs

```diff
@@ -332,6 +335,22 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> LayoutOfHelpers<'tcx> for InterpC
     }
 }
 
+impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> FnAbiOfHelpers<'tcx> for InterpCx<'mir, 'tcx, M> {
+    type FnAbiOfResult = InterpResult<'tcx, &'tcx FnAbi<'tcx, Ty<'tcx>>>;
+
+    fn handle_fn_abi_err(
+        &self,
+        err: FnAbiError<'tcx>,
+        _span: Span,
+        _fn_abi_request: FnAbiRequest<'tcx>,
+    ) -> InterpErrorInfo<'tcx> {
+        match err {
+            FnAbiError::Layout(err) => err_inval!(Layout(err)).into(),
+            FnAbiError::AdjustForForeignAbi(err) => err_inval!(FnAbi(err)).into(),
```

Comment:
You can turn this into a fatal error - this just means the current `--target` doesn't support that ABI, which should probably be a `typeck` error.

Reply:
You mean, I can ICE here?

Reply:
`rustc` has non-bug fatal errors. They're rarely used, since we want to keep going on error, but if you can't return the right value, they're an okay fit.

You can see them used for some layout errors but not others, in the existing `rustc_codegen_llvm` handling of those: https://github.com/rust-lang/rust/blob/0fb1c371d4a14f9ce7a721d8aea683a6e6774f6c/compiler/rustc_codegen_llvm/src/context.rs#L864-L865

(you could also ICE, but for now a fatal error works just as well)

Reply:
Hm, I am not sure why that is any better than an `InvalidProgram` Miri error.

Reply:
In that case can you rename the `FnAbi` case to `FnAbiAdjustForForeignAbi` or something?

Since it's not really a typical "something is wrong with the types" error but "the passed `--target` lacks FFI support" which for some reason isn't checked back in `typeck` or even earlier (well, the exact reason is that the code in question is `match`-ing on a string instead of there being any API to get some kind of typed "supported target arch" value that can be exhaustively matched on).

Reply:
Okay, rename done.